ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SourceBuilder.java

```diff
@@ -81,25 +89,29 @@ public static SourceBuilder instance() {
     final boolean forceMaterialization = !planInfo.isRepartitionedInPlan(streamSource);
 
     final KTable<K, GenericRow> transformed = source.transformValues(
-        new AddKeyAndPseudoColumns<>(keyGenerator, streamSource.getPseudoColumnVersion()));
+        new AddPseudoColumnsToMaterialize<>(streamSource.getPseudoColumnVersion()));
+
+    final KTable<K, GenericRow> toPotentiallyMaterialize;
 
     if (forceMaterialization) {
       // add this identity mapValues call to prevent the source-changelog
       // optimization in kafka streams - we don't want this optimization to
       // be enabled because we cannot require symmetric serialization between
       // producer and KSQL (see https://issues.apache.org/jira/browse/KAFKA-10179
       // and https://github.com/confluentinc/ksql/issues/5673 for more details)
-      return transformed.mapValues(row -> row, materialized);
-
+      toPotentiallyMaterialize = transformed.mapValues(row -> row, materialized);
```

Comment:
We don't need a separate mapValues() step for materialization anymore -- we can just materialize the result of the transformValues() above directly.

Reply:
If we're forcing a materialization above (outside of the `forceMaterialization` block, in the `transformValues()` step itself I assume), then should I get rid of this if/else block entirely, and make the materialization in the `transformValues()` step dependent on `forceMaterialization?

Reply:
You're still going to need the if-else block in order to decide which method variation to call on transformValues(), i.e., to decide whether to materialize or not, but you can pull logic into/out of the if-else block as needed to get this to work.